### PR TITLE
EASY-1982 check if datasetId exists before querying EMD

### PR DESCRIPTION
**Pull request by stephanHogenboom**
fixesEASY-1982 check if datasetId exists before querying EMD

#### When applied it will
*  will first check if the datasetId exist before querying the EMD
*  if it does not exist an 404 is returned 
* if the parameter ?datasetId=<id> is not provided is returns a 400

- [ ] test thoroughly on easy
- [x]  ~~does it need to be an observable~~

#### related pull requests on github
repo                       | PR
-------------------------- | -----------------
easy-                      | [PR#](PRlink)

### CODE REVIEW

File: src/test/scala/nl/knaw/dans/easy/agreement/AgreementCreatorServletSpec.scala [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+
+  private val app = new AgreementCreatorApp(minimalAppConfig)
+  private val agreementCreatorServlet = new AgreementCreatorServlet(app)
+  addServlet(agreementCreatorServlet, "/*") //same as in nl.knaw.dans.easy.agreement.AgreementCreatorService
+
+  "get /" should "respond with the service is running" in {
+    get("/") {
+      status shouldBe 200
+      body shouldBe "Agreement Creator Service running"
+    } //TODO add version once the EASY-2002 gets merged
```
Comment (rvanheest) [inline src/test/scala/nl/knaw/dans/easy/agreement/AgreementCreatorServletSpec.scala]: done 😉

File: src/main/scala/nl/knaw/dans/easy/agreement/internal/Fedora.scala [hunk 2]
```diff
@@ -111,4 +121,17 @@ case class FedoraImpl(client: FedoraClient) extends Fedora {
...
+
+  override def datasetIdExists(datasetID: DatasetID): Boolean = { //TODO does this also have to be an observable?
+    val query =
+      s"""
+         |prefix dc: <http://purl.org/dc/elements/1.1/identifier>
+         |ASK  {?s dc: "$datasetID" }
+    """.stripMargin
+    managed(Source.fromInputStream(new RiSearch(query).lang("sparql").format("csv").execute(client).getEntityInputStream))
+      .acquireAndGet(result => {
+        BooleanUtils.toBoolean(result.getLines().toList.last)
```
[the remark this answers is not in the document]
Comment (stephanHogenboom) [reply src/main/scala/nl/knaw/dans/easy/agreement/internal/Fedora.scala]: I think so,  unless a connectio or IO error occurs it always returns the headers as first value and `true` if the triple exist or `false` if it does not exist.

File: src/main/scala/nl/knaw/dans/easy/agreement/AgreementCreatorServlet.scala [hunk 3]
```diff
@@ -17,43 +17,62 @@ package nl.knaw.dans.easy.agreement
...
-    var success = false // TODO: get rid of var
-    val output = new ByteArrayOutputStream()
+  private def createAgreement(parameters: Params, output: ByteArrayOutputStream): Try[Unit] = Try {
     output
       .usedIn(AgreementCreator(parameters).createAgreement)
-      .doOnCompleted {
-        parameters.close()
-        success = true
-      }
+      .doOnCompleted(parameters.close())
```
[the remark this answers is not in the document]
Comment (stephanHogenboom) [reply src/main/scala/nl/knaw/dans/easy/agreement/AgreementCreatorServlet.scala]: I will look into it. I am not that familiar with observables, as far as i  have tested it, it seems to work good.

File: src/main/scala/nl/knaw/dans/easy/agreement/AgreementCreatorServlet.scala [hunk 4]
```diff
@@ -17,43 +17,68 @@ package nl.knaw.dans.easy.agreement
...
+      case Failure(nse: NoSuchElementException) => NotFound(nse.getMessage)
+      case Failure(iae: IllegalArgumentException) => BadRequest(iae.getMessage)
+      case Failure(t: Throwable) => InternalServerError(t.getMessage)
+    }).logResponse
+  }
 
-    val parameters = new Params(
+  private def validateDatasetIdExistsInFedora(pars: Params): Try[Unit] = {
+    pars.fedora.datasetIdExists(pars.datasetID)
+      .flatMap(if (_) Success(())
```
Comment (rvanheest) [inline src/main/scala/nl/knaw/dans/easy/agreement/AgreementCreatorServlet.scala]: would it be clearer using pattern matching?

```scala
xs.flatMap {
  case true => ...
  case false => ...
}
```

File: src/main/scala/nl/knaw/dans/easy/agreement/Command.scala [hunk 5]
```diff
@@ -41,42 +41,55 @@ object Command extends App with DebugEnhancedLogging {
...
-          .toBlocking
-          .subscribe(
-            _ => {},
-            e => logger.error("An error was caught in main:", e),
-            () => logger.debug("completed"))
-        if (success) Success(s"Created agreement for ${ commandLine.datasetID() }")
-        else Failure(new Exception(s"Could not create agreement for ${ commandLine.datasetID() }"))
+  private def validateDatasetIdExists(parameters: internal.Parameters): Try[Unit] = {
+    parameters.fedora.datasetIdExists(parameters.datasetID)
+      .flatMap(if (_) Success(())
```
Comment (rvanheest) [inline src/main/scala/nl/knaw/dans/easy/agreement/Command.scala]: same here...

File: src/main/scala/nl/knaw/dans/easy/agreement/AgreementCreatorServlet.scala [hunk 6]
```diff
@@ -17,43 +17,69 @@ package nl.knaw.dans.easy.agreement
...
+
+    val result = for {
+      params <- createParameters()
+        .recoverWith { case nse: NoSuchElementException => Failure(new IllegalArgumentException(s"mandatory parameter was not provided: ${ nse.getMessage }")) }
+      _ <- validateDatasetIdExistsInFedora(params)
+      _ <- createAgreement(params, output)
+    } yield ()
+
+    (result match {
+      case Success(_) => Ok(output.toByteArray.toString)
```
Comment (rvanheest) [inline src/main/scala/nl/knaw/dans/easy/agreement/AgreementCreatorServlet.scala]:
```suggestion
      case Success(_) => Ok(output.toByteArray)
```

File: src/main/scala/nl/knaw/dans/easy/agreement/AgreementCreatorServlet.scala [hunk 7]
```diff
@@ -17,43 +17,69 @@ package nl.knaw.dans.easy.agreement
...
+      case Failure(iae: IllegalArgumentException) => BadRequest(iae.getMessage)
+      case Failure(t: Throwable) => InternalServerError(t.getMessage)
+    }).logResponse
+  }
 
-    val parameters = new Params(
+  private def validateDatasetIdExistsInFedora(pars: Params): Try[Unit] = {
+    pars.fedora.datasetIdExists(pars.datasetID).flatMap {
+      case true => Success(())
+      case false => Failure(new NoSuchElementException(s"Dataset ID ${ pars.datasetID } was not found in fedora"))
```
Comment (rvanheest) [inline src/main/scala/nl/knaw/dans/easy/agreement/AgreementCreatorServlet.scala]:
```suggestion
      case false => Failure(new NoSuchElementException(s"DatasetId ${ pars.datasetID } does not exist")))
```

File: src/main/scala/nl/knaw/dans/easy/agreement/Command.scala [hunk 8]
```diff
@@ -41,42 +41,56 @@ object Command extends App with DebugEnhancedLogging {
...
-          .subscribe(
-            _ => {},
-            e => logger.error("An error was caught in main:", e),
-            () => logger.debug("completed"))
-        if (success) Success(s"Created agreement for ${ commandLine.datasetID() }")
-        else Failure(new Exception(s"Could not create agreement for ${ commandLine.datasetID() }"))
+  private def validateDatasetIdExists(parameters: internal.Parameters): Try[Unit] = {
+    parameters.fedora.datasetIdExists(parameters.datasetID).flatMap {
+      case true => Success(())
+      case false => Failure(new IllegalArgumentException(s"datasetID ${ parameters.datasetID } does not exist in Fedora"))
```
Comment (rvanheest) [inline src/main/scala/nl/knaw/dans/easy/agreement/Command.scala]:
```suggestion
      case false => Failure(new IllegalArgumentException(s"DatasetId ${ pars.datasetID } does not exist"))
```

**Comment by rvanheest** [comment]
review in https://github.com/stephanHogenboom/easy-license-creator/pull/1

**Pull request merged**